### ISSUE TITLE
What about reusable classes?

### ISSUE THREAD

**Issue by omeshechkov** [angular/di.js#105]
Hi,

For example I use the same class among multiple projects, say it calls SuperService which depends on Repository class.

SuperService.js:

``` javascript
var SuperService = function(repository) {
...
};
```

In order to tell injector that repository argument is a dependency I should point out a concrete realisation via annotations, that is I have to use require function (I'm using NodeJS).

SuperService.js:

``` javasript
var di = require('di');
var SQLRepository = require('./sqlRepository');
...
SuperService.annotations = [new di.Inject(SQLRepository)];
```

Everything is great while I'm using SQLRepository everywhere.

But... if I want to use this service with another dependecy within only one of projects I won't be able to do this. Because I have to change SuperService file (change require function parameter).

In other words, I want to wire classes within particular application rather than class.
Can I do this?

Thanks!

**Comment by phuvo** [comment]
Firstly, you should create a `Repository` interface.

`Repository.js`

``` javascript
function Repository() {}

module.exports = Repository;
```

Then, you should create providers for your interface.

`SqlRepository.js`

``` javascript
var di = require('di');
var Repository = require('./Repository');

function SqlRepository() {}
di.annotate(SqlRepository, new di.Provide(Repository));

module.exports = SqlRepository;
```

`FileRepository.js`

``` javascript
var di = require('di');
var Repository = require('./Repository');

function FileRepository() {}
di.annotate(FileRepository, new di.Provide(Repository));

module.exports = FileRepository;
```

You should only require the interface and leave the task of choosing a concrete provider to your main application.

`SuperService.js`

``` javascript
var di = require('di');
var Repository = require('./Repository');

function SuperService(repository) {}
di.annotate(SuperService, new di.Inject(Repository));

module.exports = SuperService;
```

In an application which uses `SqlRepository`:

``` javascript
var SqlRepository = require('./SqlRepository');
var SuperService = require('./SuperService');

var injector = new Injector([SqlRepository]);

// This is an instance of SuperService which uses SqlRepository
var superService = injector.get(SuperService);
```

In another application which uses `FileRepository`:

``` javascript
var FileRepository = require('FileRepository');
var SuperService = require('./SuperService');

var injector = new Injector([FileRepository]);

// This is an instance of SuperService which use FileRepository
var superService = injector.get(SuperService);
```

**Comment by omeshechkov** [comment]
Wow. Awesome! :)
